cpp/Naive.cpp

```diff
@@ -1,65 +1,413 @@
+        fft(aParEq, aParEq_K);
 
+        // DEBUG
+        print(aParEq);
+        print(aParEq_K);
+        // Transform moments into phase space
         for (int m = 0; m < M; ++m) {
-            plan(getMomentSlice(moments, m), getMomentSlice(momentsPH, m));
+            if (m == A_PAR) continue;
+            for_each_xy([&](Dim x, Dim y) {
+                moments_K(m, x, y) = 0;
+            });
         }
+
+        // aPar and uekPar
+        for_each_kxky([&](Dim ky, Dim kx) {
+            aParEq_K(kx, ky) /= std::sqrt(KX*KY);
+            moments_K(A_PAR, kx, ky) = aParEq_K(kx, ky);
+            ueKPar_K(kx, ky) = kPerp(kx, ky) * moments_K(A_PAR, kx, ky);
+        });
+        print(aParEq_K);
     }
 
     void Naive::run() {
+        // store all derivatives
+        DxDy<fftw::mdbuffer<3u>> dGM{M, X, Y};
+        DxDy<fftw::mdbuffer<2u>> dPhi{X, Y}, dUEKPar{X, Y};
+
+        bool divergent = false, repeat = false, noInc = false;
+        Real dt{0};
+        HyperCoefficients hyper{};
 
         for (int t = 0; t < N; ++t) {
             // predictor step
+            derivatives(phi_K, dPhi);
+            derivatives(ueKPar_K, dPhi);
             for (int m = 0; m < M; ++m) {
-// TODO
-//                plan(getMomentSlice(m), getMomentSlice());
+                derivatives(sliceXY(moments_K, m), sliceXY(dGM, m));
+            }
+
+            // DEBUG
+            print(sliceXY(moments_K, A_PAR));
+
+            if (repeat or divergent) {
+                t--; // repeat previous timestep
+            } else {
+                dt = getTimestep(dPhi, sliceXY(dGM, N_E), sliceXY(dGM, A_PAR));
+                hyper = HyperCoefficients::calculate(dt, KX, KY, M);
+            }
+
+            // DEBUG
+            std::cout << "dt: " << dt << std::endl;
+
+            // store results of nonlinear operators, as well as results of predictor step
+            fftw::mdbuffer<3u> GM_K_Star{M, X, Y}, GM_Nonlinear_K{M, X, Y};
+
+            // Compute N
+            auto bracketPhiNE_K = halfBracket(dPhi, sliceXY(dGM, N_E));
+            auto bracketAParUEKPar_K = halfBracket(sliceXY(dGM, A_PAR), dUEKPar);
+
+            // Compute A
+            DxDy<fftw::mdbuffer<2u>> dPhiNeG2{X, Y};
+            for_each_xy([&](Dim x, Dim y) {
+                dPhiNeG2.DX(x, y) =
+                        dPhi.DX(x, y) - rhoS * rhoS * (std::sqrt(2) * dGM.DX(G_MIN, x, y) + dGM.DX(N_E, x, y));
+                dPhiNeG2.DY(x, y) =
+                        dPhi.DY(x, y) - rhoS * rhoS * (std::sqrt(2) * dGM.DY(G_MIN, x, y) + dGM.DY(N_E, x, y));
+            });
+
+            auto bracketAParPhiG2Ne_K = halfBracket(sliceXY(dGM, A_PAR), dPhiNeG2);
+            auto bracketPhiDeUEKPar_K = halfBracket(dPhi, dUEKPar);
+
+            // Compute G2
+            auto bracketPhiG2_K = halfBracket(dPhi, sliceXY(dGM, G_MIN));
+            auto bracketAParG3_K = halfBracket(sliceXY(dGM, A_PAR), sliceXY(dGM, G_MIN + 1));
+
+            // Compute G_{M-1}
```

Comment:
why is the last M-1 and not M?

Reply:
I made them 0-indexed, but this can easily be worked around by passing a different `M` on the command line